runners/spark/src/test/java/org/apache/beam/runners/spark/SparkPortableExecutionTest.java

```diff
@@ -163,4 +168,75 @@ public void process(ProcessContext context) {
       Thread.sleep(1000);
     }
   }
+
+  /**
+   * Verifies that each executable stage runs exactly once, even if that executable stage has
+   * multiple outputs. While re-computation may be necessary in the event of failure, re-computation
+   * of a whole executable stage is expensive and can cause unexpected behavior when the executable
+   * stage has side effects (BEAM-7131).
+   */
+  @Test(timeout = 120_000)
+  public void testParDoWithSideEffects() throws Exception {
+    PipelineOptions options = PipelineOptionsFactory.create();
+    options.setRunner(CrashingRunner.class);
+    options
+        .as(PortablePipelineOptions.class)
+        .setDefaultEnvironmentType(Environments.ENVIRONMENT_EMBEDDED);
+    Pipeline pipeline = Pipeline.create(options);
+    Path path = FileSystems.getDefault().getPath(TEST_DIR_PREFIX, UUID.randomUUID().toString());
```

Comment:
We can use TemporaryFolder to avoid cleanup later

Example https://github.com/apache/beam/blob/b953645ed6db837d24284d7fe1fe091e7309f821/sdks/java/core/src/test/java/org/apache/beam/sdk/io/LocalFileSystemTest.java#L61